import offers during model import

kian99 opened the pull request:
## Description

This PR does two things:
1. Refactors the ImportModel method. I started getting linting errors on cognitive complexity when trying to add functionality so I took the time to refactor the method.
2. Adds functionality to the ImportModel method to import a model's application-offers.

**Please Note:** The first commit contains only the first change and the second commit contains the second change.

One issue that remains is how to handle offer URLs that already exist. This will be tackled in a separate PR.

Partially fixes [JUJU-7011](https://warthogs.atlassian.net/browse/JUJU-7011)

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

[JUJU-7011]: https://warthogs.atlassian.net/browse/JUJU-7011?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
refactor import model function

- drive by fix to openfga rebac test that incorrectly assumed ordering

alesstimec commented on internal/jimm/controller.go:
```diff
@@ -505,11 +514,28 @@ func (j *JIMM) ImportModel(ctx context.Context, user *openfga.User, controllerNa
 		}
+		return nil
+	})
+	if err != nil {
```
question: we run handleModelDeltas on L516, which will connect to the controller, create an all watcher and process a single set of deltas.. is that the intent?

kian99 replied:
I can't say why it's like that, but that does seem intentional and has been that way for some time, the tests also verify the delta functionality.
The tests show that the deltas are used to update some properties of the model like the status fields, etc.

alesstimec commented on internal/jimm/controller.go:
```diff
@@ -505,11 +514,28 @@ func (j *JIMM) ImportModel(ctx context.Context, user *openfga.User, controllerNa
+			return errors.E(op, err)
+		}
+		for _, offer := range offersToAdd {
+			var dbOffer dbmodel.ApplicationOffer
```
while this will add offers to JIMM's database, it will not grant anybody access to imported offers. At the very least, the user importing the model should be granted access.

this will require a bit of thought: each offer comes with a set of users that have access to the offer. If those are external users (those that JIMM cares about), they should be granted access to the offer when it is imported. But i guess in most cases these users would be local users - access to the offer will remain granted on the controller, but JIMM will not know/care about those rights.  perhaps we should simply print this info and let the user importing the model handle it. e.g. we imported this model and these app offer and these are the users that have access to the app offer - handle it!

kian99 replied:
True, I think we need to create an OpenFGA relation between the offers and the model. That way the model owner will automatically get access. Then they can grant access to other users. The users on the controller should never (at least with the new JIMM) be external users, because nowadays those are never stored on Juju's end (only in JIMM) so I think we can safely discard the users coming from the controller.

But yes printing a message would be helpful.

ale8k replied:
+1, relation makes sense to me.

ale8k replied:
Can do in a follow up PR @alesstimec

alesstimec reviewed (changes requested):
We need to discuss how to set access rights to imported application offers.

kian99 committed:
import offers during model import

- ensure model owner has access to the offer

SimoneDutto reviewed (approved):
lgtm

alesstimec commented on internal/jimm/controller.go:
```diff
@@ -393,133 +393,231 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
 
-	if err := j.checkJimmAdmin(user); err != nil {
-		return err
+func newModelImporter(jimm *JIMM) modelImporter {
```
pass in the new owner name so that you can immediately return an error if we fail to parse the owner tag or owner is a local user
```
func newModelImporter(jimm *JIMM, ownerTag names.UserTag) (error, *modelImporter)

alesstimec commented on internal/dbmodel/model.go:
```diff
@@ -103,7 +103,7 @@ func (m *Model) SetTag(t names.ModelTag) {
 }
 
-// FromModelUpdate updates the model from the given ModelUpdate.
+// SwitchOwner updates the model owner.
```
nitpick: SetOwner seems a more appropriate name

alesstimec commented on internal/jimm/controller.go:
```diff
@@ -393,133 +393,231 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
+
+	// TODO(CSS-5458): Remove the below section on cloud credentials once we no longer persist the relation between
+	// cloud credentials and models.
+	// Update: We need to investigate this further, if a user updates their cloud-credential it will update the credential
```
hm. this might be a bit confusing. if a user updates their cloud-credential, sure, jimm will try and update the cloud credential on the controller responsible for this model, but the actual credential used by the model will not be affected. the user would need to use the `juju set-credential` command to set credentials first and then update them to get the actual model cloud credentials updated.

kian99 replied:
Yes this will need fixing I think. For a follow up PR.

ale8k commented on internal/jimm/controller.go:
```diff
@@ -393,133 +393,231 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
+	})
+}
+
+// ImportModel imports model with the specified UUID from the controller.
```
imports a model and the currently existing offers into JIMM. A new owner must be set to represent the external user who will own this model.

Something like this for the comment?

ale8k reviewed (approved):
lgtm

kian99 committed:
PR comments

- Rename SwitchOwner to SetOwner
- Updated newModelImporter constructor
- Updated godoc

alesstimec commented on internal/jimm/controller.go:
```diff
@@ -393,133 +393,244 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
+
+	m.originalOwner, err = names.ParseUserTag(m.modelInfo.OwnerTag)
+	if err != nil {
+		return errors.E(fmt.Sprintf("invalid username %s from original model owner", m.modelInfo.OwnerTag))
```
nothing to fix, just to say that it would be very unusual to face this error

kian99 replied:
Yeah...

alesstimec commented on internal/jimm/controller.go:
```diff
@@ -393,133 +393,244 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
+	if !names.IsValidUser(newOwner) {
+		return modelImporter, errors.E(errors.CodeBadRequest, "invalid new username for new model owner")
 	}
+	newOwnerTag := names.NewUserTag(newOwner)
```
can we also check if this is a local user here?

kian99 replied:
We already do the check when we switch the owner, so I think it's okay like this.

alesstimec commented on internal/jimm/controller.go:
```diff
@@ -393,133 +393,244 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
-	// TODO(CSS-5458): Remove the below section on cloud credentials once we no longer persist the relation between
-	// cloud credentials and models
+	for _, offer := range m.offersToAdd {
+		err := m.jimm.OpenFGAClient.AddModelApplicationOffer(ctx, m.model.ResourceTag(), names.NewApplicationOfferTag(offer.OfferUUID))
```
maybe just a short godoc explaining that by associating offer with a model we are giving the model owner admin access to the offer

alesstimec reviewed (approved):
LGTM with a few suggestions

kian99 committed:
add godoc to modelImport.addPermissions
